Fix InvalidURIError with git ssh remotes

grim-luminal opened the pull request:
Replace addressable, which doesn't support ssh/git style URIs with a
combination of URI and URI::GitSsh to avoid errors when detecting the
slug from remotes.

Fixes #342

grim-luminal commented:
Looks like there's still issues to work out with Ruby 1.8.7 here.

grim-luminal commented:
The build is now passing for ruby 1.8.7 (as well as everything else).

simplyianm commented:
:+1:

 this error is annoying

BanzaiMan commented on travis.gemspec:
```diff
@@ -299,7 +299,9 @@ Gem::Specification.new do |s|
   s.add_dependency "typhoeus",              "~> 0.6", ">= 0.6.8"
   s.add_dependency "pusher-client",         "~> 0.4"
-  s.add_dependency "addressable",           "~> 2.3"
+  if Gem::Version.new(RUBY_VERSION) < Gem::Version.new("1.9.0")
```
This doesn't seem necessary, as we require Ruby 1.9.3 now.

BanzaiMan replied:
Ah, I see. We are still running builds on 1.8.7. We should probably fix that.

grim-luminal replied:
>  Ah, I see. We are still running builds on 1.8.7. We should probably fix that.

Nice. 1.9.3 or greater makes things much easier. I can add that change in this PR if it makes sense (any other PR is going to fail the build until the issue with addressable is resolved anyways).

BanzaiMan replied:
That's already taken care of on the `master` branch. You'd have to update this PR.

grim-luminal replied:
Thanks, this should be fixed now.

BanzaiMan commented on lib/travis/cli/repo_command.rb:
```diff
@@ -83,6 +82,14 @@ def detect_slug
 
+        def parse_remote(url)
+          URI.parse(url).path
+        rescue URI::InvalidURIError
```
I am against using exceptions as a part of the flow control. If `git@…` is an expected "URI", we should handle it as such.

grim-luminal replied:
This has been updated to explicitly handle `git@github.com` remotes instead of relying on the `InvalidURIError`.

BanzaiMan commented:
Also, there should be specs.

grim-luminal commented:
> Also, there should be specs.

I totally agree. I was looking for a good way to do that, but as things stand this would require writing specs for a private method, which I can do, but it seems somewhat odd. I did look into extracting the Git remote URI handling code out to it's own module under `Travis::Tools` and writing specs for that but I wasn't sure if Tools was the right place to put it.

grim-luminal commented:
@BanzaiMan I added some specs and addressed your other concerns. The specs seem weird to me and I'm totally open to doing this a different way if you any suggestions as to where to move the remote parsing stuff to make it easier to test.

BanzaiMan commented on lib/travis/cli/repo_command.rb:
```diff
@@ -83,6 +82,16 @@ def detect_slug
         end
 
+        def parse_remote(url)
+          if url.start_with?('git@github.com:')
```
I just noticed that this assumes that the remote repo resides on github.com. This is true for the most part, but we also have Travis Enterprise, which works with GitHub Enterprise. From what I understand, the only thing we know for certain is that the user name is `git`.

Perhaps
```ruby
if url =~ /^git@[^:]+:/
⋮
```
is reasonably close (if not completely accurate).

grim-luminal commented:
Updated to handle Github Enterprise URIs.

BanzaiMan commented:
OK. I will release a 1.8.1 tomorrow.
